### PR TITLE
Carousel Component

### PR DESCRIPTION
**Pull request by codeswan**
Rewrote carousel component to work properly.  Also includes options such as infinite scroll, persistent navigation, autoplay, and autoplay speed, all of which the user can customize if desired.

### CODE REVIEW

File: src/js/components/Carousel.js [hunk 1]
```diff
@@ -1,45 +1,174 @@
...
+    window.addEventListener('resize', this._onWindowResize);
+  },
+
+  componentWillUnmount: function() {
+    clearInterval(this.slideAnimation);
+
+    window.removeEventListener('resize', this._onWindowResize);
+  },
+
+  slideAnimation: null,
```
Comment (ericsoderberghp) [inline src/js/components/Carousel.js]: Can we prefix with '_' since this is internal to the module?

File: src/js/components/Carousel.js [hunk 2]
```diff
@@ -1,45 +1,174 @@
...
+    var numSlides = this.props.children.length;
+    this.setState({
+      activeIndex: (this.state.activeIndex + 1) % numSlides
+    });
+  },
+
+  _renderPrevButton: function() {
+    if (this.props.infinite || this.state.activeIndex !== 0) {
+      return (
+        <svg className={CLASS_ROOT + '__arrow ' + CLASS_ROOT + '__arrow--prev'} viewBox="0 0 36 72" version="1.1"
```
Comment (ericsoderberghp) [inline src/js/components/Carousel.js]: Is it possible to re-use the Previous and Next icons we already have?

File: src/js/components/Carousel.js [hunk 3]
```diff
@@ -1,45 +1,174 @@
...
+          <line x1="36" y1="36" x2="0" y2="72" />
+        </svg>
+      );
+    }
+  },
+
   render: function () {
     var classes = [CLASS_ROOT];
-    classes.push(CLASS_ROOT + "--toggle-" + (this.state.sequence % 2));
     if (this.props.className) {
```
Comment (ericsoderberghp) [inline src/js/components/Carousel.js]: This is extremely minor, but our convention is to put any passed in className as the last class. It makes it slightly easier to read the class names in the browser developer tools when they are grouped that way.

File: src/scss/grommet-core/_objects.carousel.scss [hunk 4]
```diff
@@ -5,79 +5,83 @@
...
+  &__arrow {
+    @include animation('carousel-reveal 1s');
+    z-index: 1;
+    width: $inuit-base-spacing-unit * 1.5;
+    height: $inuit-base-spacing-unit * 3;
+    position: absolute;
+    top: 50%;
+    transform: translateY(-50%);
+    cursor: pointer;
+    filter: drop-shadow(1px 1px 1px rgba(170, 170, 170, 0.5));
```
Comment (ericsoderberghp) [inline src/scss/grommet-core/_objects.carousel.scss]: Can we make the color a variable and put it in _settings.defaults.scss. That will allow us to reuse it and to make it easier for people theming their app to change it if needed.

**Comment by ericsoderberghp** [comment]
Nicely done. Just a few minor comments. Also, can you update CarouselDoc in grommet-docs?

**Comment by codeswan** [comment]
Thanks for the feedback @ericsoderberghp, I've made the updates you requested.  Yep, I'll make the updates to CarouselDoc and submit a PR for that as soon as it's ready.

**Pull request merged**